### PR TITLE
[bytecode runtime] Rework fixed register assignments on m68k

### PR DESCRIPTION
**Pull request by ghost**
As reported in #13249, the fixed use of %a5 in the bytecode interpreter conflicts with its internal usage as a GOT base by Glibc, and given the scarce number of available registers on this platform, it is probably better to trust the compiler to perform register assignment here.

### CODE REVIEW

**Comment by gasche** [comment]
This PR appears to fix a bug that would let Debian testing progress further, and in this sense it is good.

On the other hand I am not sure about the idea of removing the register assignment because there are too few registers. Even when there are few, quite probably the PC, SP and ACCU registers of the VM are those you really want to have in machine registers. Has this calculus changed in OCaml 5 rather than OCaml 4? Could you suggest other registers to reserve instead? (Do we even have a reliable way to do a performance comparison of the two approaches?)

**Comment by ghost** [comment]
> This PR appears to fix a bug that would let Debian testing progress further, and in this sense it is good.
> 
> On the other hand I am not sure about the idea of removing the register assignment because there are too few registers. Even when there are few, quite probably the PC, SP and ACCU registers of the VM are those you really want to have in machine registers. Has this calculus changed in OCaml 5 rather than OCaml 4? Could you suggest other registers to reserve instead? (Do we even have a reliable way to do a performance comparison of the two approaches?)

This has not changed from OCaml 4 and I expect OCaml 4 to fail the same way - at least when the interpreter needs to invoke a libc function which accesses global data (and thus will perform `%a5`-related loads).

On m68k, register allocation is painful, because although there are 16 general-purpose registers, they are not all equal, only `%a0`-`%a7` can be used for addressing (e.g. for register-related memory loads and stores), and only `%d0`-`%d7` can be used for data computation. Also, `%a7` is the stack pointer and `%a6` the frame pointer. Now that `%a5` is also reserved as the GOT pointer, this leaves only five address registers; at this point I think it is better to trust the compiler register allocation among these five registers, rather than force two of them to be fixed and only three left.

It might be worth trying to keep fixed allocations but use `%a3` instead of `%a5` for `PC_REG` and compare execution times against the "no fixed assignment" version suggested in this PR, though.

**Comment by gasche** [comment]
I would be more at ease with the minimal/conservative change (using a3 instead of a5) first. We could wait for a performance test (which probably is never going to come because few people care about this architecture) to remove the assignments altogether.

**Comment by ghost** [comment]
> I would be more at ease with the minimal/conservative change (using a3 instead of a5) first. We could wait for a performance test (which probably is never going to come because few people care about this architecture) to remove the assignments altogether.

How about this new version?

**Comment by gasche** [comment]
I feel happy with this new version -- as in, I don't have the expertise to ensure that this is right, but my risk assessment is that this is very probably fine if you say it is, so we could go ahead.

What are the benefits of keeping a5 on non-linux systems, instead of switching to a3 uniformly, which seems simpler?

@glondu if you have an easy way to check that this patch works just as well as disabling the register assignment, I would be interested in a confirmation. But I am thinking of approving&merging it anyway.

**Comment by ghost** [comment]
> What are the benefits of keeping a5 on non-linux systems, instead of switching to a3 uniformly, which seems simpler?

I don't know. If someone is willing to dig into history, the rationale behind the choice of the existing assignment might provide useful information (as in: "we tried various combinations and this one turns out to be better because...") but I'm not sure there are traces of it.

**Review by gasche — APPROVED** [review]
Hm. I don't understand enough about the domain to see why a5->a3 would make a difference, but I don't want to insist and I am happy to approve the PR as-is.

**Comment by gasche** [comment]
I looked at the git history, the m68k register assignment was chosen in 1996 by @xavierleroy and never changed since: 9a374eb4d60a489df981de8e9e68ddf100e597bf.

**Comment by ghost** [comment]
Yet this proves there were still Sun3 running in INRIA in 1996! At least in Rocquencourt - INRIA Sophia-Antipolis was already full SPARC (and Linux) at that time. :grinning:

**Comment by glondu** [comment]
> @glondu if you have an easy way to check that this patch works just as well as disabling the register assignment, I would be interested in a confirmation. But I am thinking of approving&merging it anyway.

I've just uploaded to experimental a new version of the package, with the current state of this PR + disabling runtime_events on armel. Results should appear in a few hours there: https://buildd.debian.org/status/package.php?p=ocaml&suite=experimental

File: runtime/interp.c [hunk 1]
```diff
@@ -210,11 +210,18 @@ Caml_inline void check_trap_barrier_for_effect
 #define SP_REG asm("%r17")
 #define ACCU_REG asm("%r16")
 #endif
-#ifdef __mc68000__
+#ifdef __mc68000__ /* { */
+#ifdef __linux__
```
Comment (MisterDA) [inline runtime/interp.c]: Should this rather test for the libc? If it's glibc-specific we could try `__GNU_LIBRARY__` or `__GLIBC__` or `__GLIBC_MINOR__`. What does e.g., musl do?

Comment (ghost) [reply runtime/interp.c]: I considered this, but there is always the risk of mixing objects compiled against glibc with objects compiled against musl (which does not reserve `%a5`), so I opted for a more conservative check.

File: runtime/interp.c [hunk 2]
```diff
@@ -210,11 +210,18 @@ Caml_inline void check_trap_barrier_for_effect
...
+/* glibc relies upon %a5 to always contain the base address of the GOT. */
+#define PC_REG asm("a3")
+#define SP_REG asm("a4")
+#define ACCU_REG asm("d7")
+#else
 #define PC_REG asm("a5")
 #define SP_REG asm("a4")
 #define ACCU_REG asm("d7")
 #endif
+#endif /* } __mc68000 */
```
Comment (xavierleroy) [inline runtime/interp.c]: No need to add more ifdefs, the revised register assignment will work everywhere.
```suggestion
#ifdef __mc68000__
#define PC_REG asm("a3")
#define SP_REG asm("a4")
#define ACCU_REG asm("d7")
#endif
```

**Review by xavierleroy — COMMENTED** [review]
So there's still one M68k machine compiling Debian packages...  (I doubt anyone uses those packages, esp. the OCaml ones, but that's the way it is).  Proposed simplification below.

**Comment by xavierleroy** [comment]
> Yet this proves there were still Sun3 running in INRIA in 1996! At least in Rocquencourt

I probably restarted an old Sun 3/60 that was lying around  just to do this M68k port.  In retrospect, I should not have.

File: runtime/interp.c (hunk 2 above)
Comment (ghost) [reply runtime/interp.c]: Ok then.

**Pull request merged**